internal/provider/providers/transip/provider.go

```diff
@@ -0,0 +1,285 @@
+	"github.com/qdm12/ddns-updater/internal/provider/errors"
+	"github.com/qdm12/ddns-updater/internal/provider/headers"
+	"github.com/qdm12/ddns-updater/internal/provider/utils"
+	"github.com/qdm12/ddns-updater/pkg/publicip/ipversion"
+	"net/http"
+	"net/netip"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type Provider struct {
+	domain     string
+	owner      string
+	ipVersion  ipversion.IPVersion
+	ipv6Suffix netip.Prefix
+	username   string
+	key        string
+}
+
+func New(data json.RawMessage, domain, owner string,
+	ipVersion ipversion.IPVersion, ipv6Suffix netip.Prefix) (
+	p *Provider, err error,
+) {
+	extraSettings := struct {
+		Username string `json:"username"`
+		Key      string `json:"key"`
+	}{}
+	err = json.Unmarshal(data, &extraSettings)
+	if err != nil {
+		return nil, err
+	}
+
+	err = validateSettings(domain, extraSettings.Username, extraSettings.Key)
+	if err != nil {
+		return nil, fmt.Errorf("validating provider specific settings: %w", err)
+	}
+
+	return &Provider{
+		domain:     domain,
+		owner:      owner,
+		ipVersion:  ipVersion,
+		ipv6Suffix: ipv6Suffix,
+		username:   extraSettings.Username,
+		key:        extraSettings.Key,
+	}, nil
+}
+
+func validateSettings(domain, username string, key string) (err error) {
+	err = utils.CheckDomain(domain)
+	if err != nil {
+		return fmt.Errorf("%w: %w", errors.ErrDomainNotValid, err)
+	}
+
+	if username == "" {
+		return errors.ErrUsernameNotSet
+	}
+
+	if key == "" {
+		return errors.ErrKeyNotSet
+	}
+
+	if _, err := parsePrivateKey(key); err != nil {
+		return fmt.Errorf("%w: %w", errors.ErrKeyNotValid, err)
+	}
+
+	return nil
+}
+
+func (p *Provider) String() string {
+	return utils.ToString(p.domain, p.owner, constants.TransIP, p.ipVersion)
+}
+
+func (p *Provider) Domain() string {
+	return p.domain
+}
+
+func (p *Provider) Owner() string {
+	return p.owner
+}
+
+func (p *Provider) IPVersion() ipversion.IPVersion {
+	return p.ipVersion
+}
+
+func (p *Provider) IPv6Suffix() netip.Prefix {
+	return p.ipv6Suffix
+}
+
+func (p *Provider) Proxied() bool {
+	return false
+}
+
+func (p *Provider) BuildDomainName() string {
+	return utils.BuildDomainName(p.owner, p.domain)
+}
+
+func (p *Provider) HTML() models.HTMLRow {
+	return models.HTMLRow{
+		Domain:    fmt.Sprintf("<a href=\"http://%s\">%s</a>", p.BuildDomainName(), p.BuildDomainName()),
+		Owner:     p.Owner(),
+		Provider:  "<a href=\"https://www.transip.nl/\">TransIP</a>",
+		IPVersion: p.ipVersion.String(),
+	}
+}
+
+func parsePrivateKey(keyString string) (*rsa.PrivateKey, error) {
+	// Remove the begin and end markers, remove whitespace, and trim.
+	pemData := strings.ReplaceAll(keyString, "\n", "")
+	pemData = strings.ReplaceAll(pemData, "-----BEGIN PRIVATE KEY-----", "")
+	pemData = strings.ReplaceAll(pemData, "-----END PRIVATE KEY-----", "")
+	pemData = strings.TrimSpace(pemData)
+
+	decodedKey, err := base64.StdEncoding.DecodeString(pemData)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Would it perhaps be better to let the `encoding/pem` package do this? That makes the parsing more compliant with the PEM standard.

```suggestion
	block, _ := pem.Decode(pubPEMData)
	if block == nil || block.Type != "PRIVATE KEY" {
		return nil, errors.New("failed to decode PEM block containing private key")
	}
```

And below substitute `decodedKey` for `block.Bytes`.